optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java

```diff
@@ -114,9 +130,8 @@ protected Constraint periodPenaltyAfter(ConstraintFactory constraintFactory) {
                         filtering((periodPenalty, leftExam) -> leftExam.getPeriod() != null))
                 .join(Exam.class,
                         equal((periodPenalty, leftExam) -> periodPenalty.getRightTopic(), Exam::getTopic),
-                        filtering((periodPenalty, leftExam, rightExam) -> rightExam.getPeriod() != null),
-                        filtering((periodPenalty, leftExam,
-                                rightExam) -> leftExam.getPeriod().getPeriodIndex() <= rightExam.getPeriod().getPeriodIndex()))
+                        lessThanOrEqual((periodPenalty, leftExam) -> leftExam.getPeriodIndex(), Exam::getPeriodIndex),
+                        filtering((periodPenalty, leftExam, rightExam) -> rightExam.getPeriod() != null))
```

Comment:
Doesn't this crash? rightExam.getPeriodIndex() can throw an NPE, right? Or return null that isn't Comparable?

Ideally, the rightExam period != null filtering must happen before the join:

```
.join(f.from(Exam).filter(exam.getPeriod != null)),
       lessThanOrEqual(....)
```

Reply:
`getPeriodIndex()` has a built-in `null` check. That's how I can afford to do this.
But the `null` check here is redundant anyway, since `lessThanOrEqual()` would filter those situations out anyway.